ServiceAccountIssuerDiscovery: Add user facing documentation (#19328)
kubernetes/website: content/en/docs/tasks/configure-pod-container/configure-service-account.md

The application is responsible for reloading the token when it rotates. Periodic reloading (e.g. once every 5 minutes) is sufficient for most usecases.

## Service Account Issuer Discovery

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

{{< note >}}

> The Service Account Issuer Discovery feature is __alpha__ in 1.18 and enabled by

Reviewer: This is repeating the feature-state line. I would tell readers how to enable the feature gate (what component?) and omit the repetition.

Author: Done.

> setting the `ServiceAccountIssuerDiscovery` feature gate to `true` and then

Reviewer: ```suggestion
enabling the `ServiceAccountIssuerDiscovery` [feature gate](/docs/reference/command-line-tools-reference/feature-and then
```

Author: Done.

enabling the Service Account Token Projection feature as described above.

Note that the issuer URL must comply with the
[OIDC Discovery Spec](https://openid.net/specs/openid-connect-discovery-1_0.html). In
practice, this means it must use the `https` scheme, and should serve an OpenID

> provider configuration at `{service-account-issuer}/.well-known/openid-configuration`.

Reviewer: I'd start the ` {{< note >}}` here (move the advice about enabling the feature gate to be before the note block).

Author: Done.

If the URL does not comply, the `ServiceAccountIssuerDiscovery` endpoints will
not be registered, even if the feature is enabled.
{{< /note >}}

The Service Account Issuer Discovery feature enables federation of Kubernetes
service account tokens issued by a cluster (the _identity provider_) with
external systems (_relying parties_).

> When enabled, the Kubernetes API server will serve an OpenID Provider

Reviewer: ```suggestion
When enabled, the Kubernetes API server provides a public OpenID Provider
```

Author: Done. Also added clarification on RBAC, thanks for the reminder with your comment about "public."

Configuration document at `/.well-known/openid-configuration` and the associated
JSON Web Key Set (JWKS) at `/openid/v1/jwks`. The OpenID Provider Configuration
is sometimes referred to as the _discovery document_.

{{< note >}}
The responses served at `/.well-known/openid-configuration` and
`/openid/v1/jwks` are designed to be OIDC compatible, but not strictly OIDC

> compliant. Today they contain only the parameters necessary to perform

Reviewer: ```suggestion
compliant. Those documents contain only the parameters necessary to perform
```
(“today” looks a bit like a statement about the future; see https://kubernetes.io/docs/contribute/style/style-guide/#avoid-statements-about-the-future)

I would definitely remove:
```markdown
Additional, parameters may be
added if it is determined that doing so improves compatibility with common
client implementations used by relying parties.
```

That kind of detail belongs in the KEP.

Author: Yup, already in the KEP. Sounds good. Done.

validation of Kubernetes service account tokens. Additional, parameters may be
added if it is determined that doing so improves compatibility with common
client implementations used by relying parties.
{{< /note >}}

> The JWKS contains public keys that can be used to validate the Kubernetes

Reviewer: ```suggestion
The JWKS response contains public keys that can a relying party can use to validate the Kubernetes
```

Author: Done.

service account tokens. Relying parties first query for the OpenID Provider
Configuration, and use the `jwks_uri` field in the response to find the JWKS.

In many cases, Kubernetes API servers are not available on the public internet,
but public endpoints that serve cached responses from the API server can be made
available by users or service providers. In these cases, it is possible to
override the `jwks_uri` in the OpenID Provider Configuration so that it points
to the public endpoint, rather than the API server's address, by passing the
`--service-account-jwks-uri` flag to the API server. Like the issuer URL, the
JWKS URI is required to use the `https` scheme.

For additional technical details, see the
[Service Account Signing Key Retrieval KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190730-oidc-discovery.md)
and

> [OIDC Discovery Spec](https://openid.net/specs/openid-connect-discovery-1_0.html).

Reviewer: These could (I think should) move into the `whatsnext` capture.

Author: Done.

The application is responsible for reloading the token when it rotates. Periodic reloading (e.g. once every 5 minutes) is sufficient for most usecases.

## Service Account Issuer Discovery

{{< feature-state for_k8s_version="v1.18" state="alpha" >}}

The Service Account Issuer Discovery feature is enabled by enabling the
`ServiceAccountIssuerDiscovery` [feature gate](/docs/reference/command-line-tools-reference/feature)

> and then enabling the Service Account Token Projection feature as described above.

Reviewer: nit:
```suggestion
and then enabling the Service Account Token Projection feature as described [above](#service-account-token-volume-projection).
```

Author: Done.

{{< note >}}

> Note that the issuer URL must comply with the

Reviewer: ```suggestion
The issuer URL must comply with the
```
The shortcode ` {{< note >}}` already marks this as a note.

Author: Done.

[OIDC Discovery Spec](https://openid.net/specs/openid-connect-discovery-1_0.html). In
practice, this means it must use the `https` scheme, and should serve an OpenID
provider configuration at `{service-account-issuer}/.well-known/openid-configuration`.

If the URL does not comply, the `ServiceAccountIssuerDiscovery` endpoints will
not be registered, even if the feature is enabled.
{{< /note >}}

The Service Account Issuer Discovery feature enables federation of Kubernetes
service account tokens issued by a cluster (the _identity provider_) with
external systems (_relying parties_).

When enabled, the Kubernetes API server provides an OpenID Provider
Configuration document at `/.well-known/openid-configuration` and the associated
JSON Web Key Set (JWKS) at `/openid/v1/jwks`. The OpenID Provider Configuration
is sometimes referred to as the _discovery document_.

> When enabled, the cluster is also configured with a default RBAC `ClusterRole`

Reviewer: nit:
```suggestion
 When enabled, the cluster is also configured with a default RBAC ClusterRole
```
In the docs, API objects don't go inside backticks.

Author: Done.

requirements and which external systems they intend to federate with.

{{< note >}}
The responses served at `/.well-known/openid-configuration` and
`/openid/v1/jwks` are designed to be OIDC compatible, but not strictly OIDC
compliant. Those documents contain only the parameters necessary to perform
validation of Kubernetes service account tokens.
{{< /note >}}

The JWKS response contains public keys that a relying party can use to validate
the Kubernetes service account tokens. Relying parties first query for the
OpenID Provider Configuration, and use the `jwks_uri` field in the response to
find the JWKS.

In many cases, Kubernetes API servers are not available on the public internet,
but public endpoints that serve cached responses from the API server can be made
available by users or service providers. In these cases, it is possible to
override the `jwks_uri` in the OpenID Provider Configuration so that it points
to the public endpoint, rather than the API server's address, by passing the
`--service-account-jwks-uri` flag to the API server. Like the issuer URL, the
JWKS URI is required to use the `https` scheme.
{{% /capture %}}

{{% capture whatsnext %}}
See also:

> - [Cluster Admin Guide to Service Accounts](/docs/reference/access-authn-authz/service-accounts-admin/)

Reviewer: Add a blank line before this list to keep the Markdown parser happy.

Author: Done.
